docs/rfcs/XXXX_storage_backends.md

```diff
@@ -0,0 +1,248 @@
+
+### Management of External Buckets
+
+Dewrangle will not be responsible for configuration of the external buckets. This responsibility resides with the organization administrator and whoever controls their resources.
+
+## User Flows
+
+### Creating a Storage Backend
+
+Go to organiztion study page and add a new Storage Backend
+
+User provides basic information such as the bucket name and what prefix they want to keep data under
+![](https://i.ibb.co/n6Qxb6Q/Basic.png)
+
+The user will be asked to configure an IAM user within their AWS account that the API may use to interact with the specified bucket.
+![](https://i.ibb.co/ys5xc61/Create-User.png)
+
+The user will need to create a new programmatic IAM user within the AWS dashboard
+
+![](https://i.ibb.co/7bVWVFR/Screenshot-2021-08-10-at-15-47-08-IAM-Management-Console.png)
+
+The policy that is suggested or one that allows for similar access will need to be created to be attached to the new IAM user
+![](https://i.ibb.co/zQhnYT5/Screenshot-2021-08-10-at-15-51-49-IAM-Management-Console.png)
+
+The new IAM user can then be created
+![](https://i.ibb.co/zhj02jx/Screenshot-2021-08-10-at-15-52-42-IAM-Management-Console.png)
+
+In Dewrangle, the user may now paste in the credentials for the new IAM user
+![](https://i.ibb.co/yBXbjTJ/Create-User-1.png)
+
+Once this is done, Dewrangle will confirm that it has the permissions necessary to perform its operations
+
+![](https://i.ibb.co/1fk2B48/Create-User-3.png)
+
+The new storage backend will now be available for use in studies within the organization.
+
+
+### Modifying Storage Backends for Studies
+
+Study administrators may change a study's settings to use any Storage Backend that is in the organization.
+![](https://i.ibb.co/DQZTKCr/Screen-Shot-2021-08-10-at-4-16-15-PM.png)
+
+A study may only have one Document Storage Backend to read and write uploaded documents to. If this location changes after uploads have been made, Dewrangle will not be able to find old files unless they are migrated to the correct location in the new backend and updated in the application database. It may be possible to automate this in the future.
+
+[========]
+
+
+## Technical Implementation
+
+### StorageBackend Model
+
+The storage backend model
+
+```graphql
+type StorageBackend {
+  id: ID!
+  name: String!
+  bucket: String
+  prefix: String
+  organization: Organization
+  accessKey: String
+  secretKey: String
+  studies: [one-many]
+}
+```
+
+#### Secret Storage
+
+The `access_key` and `secret_key` will be stored in the application's database in an encrypted state. The encryption will be done with a KMS key provisioned specifically for Dewrangle. In order to perform actions on behalf of the IAM user and thus the stored key pair, the secrets must be decrypted first using the same KMS key.
```

Comment:
What are the pros/cons of storing it in the db vs in the application's secrets bucket where we currently store our secret key pair?

Reply:
I imagine that reading from S3 is a lot slower than from the database. We'd also materialize these files in S3 so we'd have to make sure they are pretty well administered to prevent accidental exposure. Keeping them in the database provides two layers of security, one from the db connection auth, and one from the KMS.